fission-web-server/library/Fission/Web/Server/Internal/Tasks.hs

```diff
@@ -19,3 +39,71 @@ deleteByUsername userNameTxt =
       User.getByUsername uname >>= \case
         Just (Entity userId _) -> runDB $ deleteCascade userId
         Nothing                -> error "User doesn't exist"
+
+---
+
+ensureAllPinned :: Server ()
+ensureAllPinned = do
+  dbCIDs      <- getAllDBPins
+  clusterURLs <- asks ipfsURLs
+  manager     <- asks ipfsHttpManager
+
+  forM_ clusterURLs \(IPFS.URL url) -> do
+    let clientManager = mkClientEnv manager url
+
+    liftIO (runClientM listPins $ mkClientEnv manager url) >>= \case
+      Left err ->
+        logError $ "Pin list reuqest failed: " <> displayShow err
+
+      Right remoteCIDs ->  do
+        let missingCIDs = dbCIDs List.\\ remoteCIDs
+
+        forM_ missingCIDs \cid@(CID hash) ->
+          liftIO (runClientM (IPFS.pin hash) clientManager) >>= \case
+            Left err ->
+              logError $ mconcat
+                [ "Pin failed: "
+                , " -- "
+                , displayShow url
+                , " -- "
+                , displayShow cid
+                , " -- "
+                , displayShow err
+                ]
+
+            Right _ ->
+              noop
+
+pinAllToCluster :: [CID] -> Server [(CID, ClientError)]
+pinAllToCluster cids =
+  foldM combiner mempty cids
+  where
+    combiner acc cid =
+      Cluster.pin cid >>= \case
+        Left err -> do
+          logWarn $ "Unable to pin " <> display err
+          return ((cid, err) : acc)
+
+        Right _ ->
+          return acc
+
+getAllDBPins :: Server [CID]
```

Comment:
Should we be basing this off of Events in the DB so that we get historical pins as well? 
I don't think it's necessary since they aren't referenced anywhere, but just thought I'd mention it, because this is not actually ensuring the pinning of the _entire_ system

Reply:
@dholms can you expand on that? I'm not sure I follow.

Reply:
We never unpin CIDs on, for instance, app updates or user data root updates.

So if we treated the original remote node's pin list as the "canonical" pin list, it would include some CIDs that are not included in the DB queries here. But those CIDs would be included in the Event Tables. 

However, we don't need that content for anything (and recent FS updates contain pointers to `previous` versions of the FS), so I think it's okay that we don't pin those. I just wanted to note the discrepancy.

Does that make sense?